unit-tests/src/main/scala/java/util/regex/MatcherSuite.scala

```diff
@@ -0,0 +1,279 @@
+    )
+
+    assertThrows2[IndexOutOfBoundsException](
+      end(42),
+      _.getMessage == "No group 42"
+    )
+  }
+
+  test("start/end") {
+    val m = matcher("a(\\d)(\\d)z", "_a12z_a34z_")
+    import m._
+
+    assert(find())
+    assertEquals(start, 1)
+    assertEquals(end, 5)
+
+    assert(find())
+    assertEquals(start, 6)
+    assertEquals(end, 10)
+
+    assert(!find())
+  }
+
+  test("start(name)/end(name) re2 syntax") {
+    val m = matcher(
+      "from (?P<S>.*) to (?P<D>.*)",
+      "from Montreal, Canada to Lausanne, Switzerland"
+    )
+    import m._
+
+    assert(find())
+    assertEquals(start("S"), 5)
+    assertEquals(end("S"), 21)
+
+    assertEquals(start("D"), 25)
+    assertEquals(end("D"), 46)
+
+    assertThrows2[IllegalArgumentException](
+      start("foo"),
+      _.getMessage == "No group with name <foo>"
+    )
+
+    assertThrows2[IllegalArgumentException](
+      end("foo"),
+      _.getMessage == "No group with name <foo>"
+    )
+  }
+
+  testFails("start(name)/end(name) java syntax", 620) {
+    // change pattern to java: "from (?<S>.*) to (?<D>.*)"
+    val m = matcher(
+      "from (?<S>.*) to (?<D>.*)",
+      "from Montreal, Canada to Lausanne, Switzerland"
+    )
+    import m._
+
+    assert(find())
+    assertEquals(start("S"), 5)
+    assertEquals(end("S"), 21)
+
+    assertEquals(start("D"), 25)
+    assertEquals(end("D"), 46)
+  }
+
+  test("appendReplacement/appendTail") {
+    val buf = new StringBuffer()
+
+    val m = matcher("a(\\d)(\\d)z", "_a12z_a34z_")
+    import m._
+
+    while (find()) {
+      appendReplacement(buf, "{" + group + "}")
+    }
+    appendTail(buf)
+    assertEquals(buf.toString, "_{a12z}_{a34z}_")
+  }
+
+  test("appendReplacement/appendTail with group replacement by index") {
+    val buf = new StringBuffer()
+
+    val m = matcher("a(\\d)(\\d)z", "_a12z_a34z_")
+    import m._
+
+    while (find()) {
+      appendReplacement(buf, "{$0}")
+    }
+    appendTail(buf)
+    assertEquals(buf.toString, "_{a12z}_{a34z}_")
+  }
+
+  test("appendReplacement/appendTail with group replacement by name") {
+    val buf = new StringBuffer()
+
+    val m = matcher(
+      "from (?P<S>.*) to (?P<D>.*)",
+      "from Montreal, Canada to Lausanne, Switzerland"
+    )
+    import m._
+
+    while (find()) {
+      appendReplacement(buf, "such ${S}, wow ${D}")
+    }
+    appendTail(buf)
+    assertEquals(buf.toString,
+                 "such Montreal, Canada, wow Lausanne, Switzerland")
+  }
+
+  test("reset") {
+    val m = matcher("a(\\d)(\\d)z", "_a12z_a34z_")
+    import m._
+
+    assert(find())
+    assertEquals(start, 1)
+    assertEquals(end, 5)
+
+    reset()
+
+    assert(find())
+    assertEquals(start, 1)
+    assertEquals(end, 5)
+
+    assert(find())
+    assertEquals(start, 6)
+    assertEquals(end, 10)
+
+    assert(!find())
+  }
+
+  // we don't support lookahead
+  testFails("(not supported) hasTransparentBounds/useTransparentBounds", 640) {
```

Comment:
With an opaque bound, it will not match the lookaheads. I think you can add useTransparentBounds to Scala.js.